accepted/2020/diagnostics/activity-improvements-2.md

```diff
@@ -0,0 +1,152 @@
+        public int Count { get; }
+        public IEnumerator<KeyValuePair<string, object>> GetEnumerator() { ... }
+        IEnumerator IEnumerable.GetEnumerator() { ... }
+    }
+```
+
+## Proposal Changing already exposed APIs in previous preview releases
+
+As we are introducing the ActivityAttributesCollection, we can now modify some of the exposed APIs to use this collection instead.
+
+### ActivityEvent
+
+```C#
+    // Already exposed APIs
+    public readonly struct ActivityEvent
+    {
+        public ActivityEvent(string name, IEnumerable<KeyValuePair<string, object>>? attributes)
+        public ActivityEvent(string name, DateTimeOffset timestamp, IEnumerable<KeyValuePair<string, object>>? attributes)
+    }
+
+    // change it to the following:
+    public readonly struct ActivityEvent
+    {
+        public ActivityEvent(string name, ActivityAttributesCollection? attributes)
+        public ActivityEvent(string name, DateTimeOffset timestamp, ActivityAttributesCollection? attributes)
+    }
+```
+
+### ActivityLink
+
+```C#
+    // Already exposed APIs
+    public readonly partial struct ActivityLink  : IEquatable<ActivityLink>
+    {
+        public ActivityLink(ActivityContext context, IEnumerable<KeyValuePair<string, object>>? attributes)
+    }
+
+    // change it to the following:
+    public readonly partial struct ActivityLink  : IEquatable<ActivityLink>
+    {
+        public ActivityLink(ActivityContext context, ActivityAttributesCollection? attributes)
+    }
+```
+
+## Automatic Trace Id  generation in case of null parent
+
+Sampler like probabilistic sampler can depend on the trace id for the sampling decision. Usually before creating any Activity using ActivitySource, the `ActivityListener.GetRequestedDataUsingContext` callback will get called which usually get implemented by the samplers. The parent context is sent as a parameter to this callback and the sampler gets the trace id from this parent context to perform the sampling action. When there is no parent, we send the default context which is nulling the trace id. The samplers depending on trace id are not going to be able to do the sampling at that time. The samplers even cannot generate and use a trace id because it cannot inform the Activity creation process to use the generated trace Id when creating the Activity object.
+There are a couple of proposals how to address this issue in the .NET runtime. The idea is the runtime need to generate the trace Id and send it to the samplers. If the sampler using the passed trace id and decide to sample in this case, the Activity which will get created because of that sampler decision will be using the generated trace Id in its context.
+
+The proposal here is to expose the `ActivityListener` property
+
+```c#
+    public sealed class ActivityListener : IDisposable // already existing type
+    {
+        public bool PregenerateNewRootId { get; set;} // other name suggestion: GenerateTraceIdForRootContext
```

Comment:
RootId - may get confused with Activity.RootId. 

`GenerateTraceIdForNullParent` - does this sound better? `GenerateTraceIdForRootActivity` is also okay. The term "root" span is used throughout official OT specs.

Reply:
I added your suggestion in the doc and we discuss that in the design review.